realm/realm-library/src/main/cpp/io_realm_internal_OsRealmConfig.cpp

```diff
@@ -0,0 +1,332 @@
+                               std::shared_ptr<SyncSession> session) {
+            realm::jni_util::Log::d("Callback to Java requesting token for path");
+
+            JNIEnv* env = realm::jni_util::JniUtils::get_env(true);
+
+            jstring access_token_string = (jstring)env->CallStaticObjectMethod(
+                sync_manager_class, java_bind_session_method, to_jstring(env, path.c_str()));
+            if (access_token_string) {
+                // reusing cached valid token
+                JStringAccessor access_token(env, access_token_string);
+                session->refresh_access_token(access_token, realm::util::Optional<std::string>(syncConfig.realm_url));
+            }
+        };
+
+        // Get logged in user
+        JStringAccessor user_id(env, j_user_id);
+        JStringAccessor auth_url(env, j_auth_url);
+        SyncUserIdentifier sync_user_identifier = {user_id, auth_url};
+        std::shared_ptr<SyncUser> user = SyncManager::shared().get_existing_logged_in_user(sync_user_identifier);
+        if (!user) {
+            JStringAccessor realm_auth_url(env, j_auth_url);
+            JStringAccessor refresh_token(env, j_reresh_token);
+            user = SyncManager::shared().get_user(sync_user_identifier, refresh_token);
+        }
+
+        util::Optional<std::array<char, 64>> sync_encryption_key(util::none);
+        if (!config.encryption_key.empty()) {
+            sync_encryption_key = std::array<char, 64>();
+            std::copy_n(config.encryption_key.begin(), 64, sync_encryption_key->begin());
+        }
+
+        // Disable ssl by default
+        bool sync_client_validate_ssl = false;
+        util::Optional<std::string> ssl_trust_certificate_path = util::none;
+
+        JStringAccessor realm_url(env, j_sync_realm_url);
+        config.sync_config = std::make_shared<SyncConfig>(SyncConfig{
+            user, realm_url, SyncSessionStopPolicy::Immediately, std::move(bind_handler), std::move(error_handler),
+            nullptr, sync_encryption_key, sync_client_validate_ssl, ssl_trust_certificate_path});
+    }
+    CATCH_STD()
+}
+
+JNIEXPORT void JNICALL Java_io_realm_internal_OsRealmConfig_nativeSetSyncConfigSslSettings(
+    JNIEnv* env, jclass, jlong native_ptr, jboolean sync_client_validate_ssl,
+    jstring j_sync_ssl_trust_certificate_path)
+{
+    TR_ENTER_PTR(native_ptr);
+
+    auto& config = *reinterpret_cast<Realm::Config*>(native_ptr);
+    REALM_ASSERT(config.sync_config);
+    REALM_ASSERT((sync_client_validate_ssl && j_sync_ssl_trust_certificate_path) ||
+                 (!sync_client_validate_ssl && !j_sync_ssl_trust_certificate_path));
+
+    try {
+        config.sync_config->client_validate_ssl = sync_client_validate_ssl;
+        if (j_sync_ssl_trust_certificate_path) {
+            JStringAccessor cert_path(env, j_sync_ssl_trust_certificate_path);
+            config.sync_config->ssl_trust_certificate_path = realm::util::Optional<std::string>(cert_path);
+        }
+        else {
```

Comment:
I am think about should we allow `nativeSetSyncConfigSslSettings` to be called multi-times. This `else` is mostly to make it safe in that case. I can add some assert to prevent it from being called more than once.